Enhanced the 'matmul' wrapper.

eaplatanios opened the pull request:
The `matmul` op now matches the behavior of the Python op (i.e., using `batchMatmul` whenever appropriate, and it also supports transposing either or both of its arguments.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -1505,25 +1505,59 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
-    // shared visibility is not allowed" SILVerifier error in
-    // "tests/AutoDiff/tensor_autodiff_runtime.swift".
-    return Raw.matMul(lhs, rhs, transposeA: false, transposeB: false)
+    if lhs.rank > 2 && rhs.rank > 2 {
```
How about

```swift
switch (lhs.rank, rhs.rank) {
case (3..., 3...):
  ...
case (2, 3...):
  ...
case (3..., 2):
  ...
default:
  ...
}
```

eaplatanios replied:
I didn't know you could pattern match on integers with ranges. That's cool! :)

eaplatanios commented:
@rxwei I addressed your feedback and also added a helper for adding tests. After we merge this I'll update all the other PRs to add tests using this helper.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -1505,25 +1505,59 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
 @inlinable
-@differentiable(vjp: _vjpMatmul(_:_:) where Scalar: TensorFlowFloatingPoint)
+@differentiable(vjp: _vjpMatmul(_:_:transposeA:transposeB:) where Scalar: TensorFlowFloatingPoint)
 public func matmul<Scalar: Numeric>(
```
How about calling it `matmul(_:transposed:_:transposed)` where the two `transposed`s have a default value?

eaplatanios replied:
Can you call this like `matmul(x, y)` or do the arguments need to be ordered?

rxwei replied:
Yep

eaplatanios replied:
That's neat. I just pushed this change.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -1525,30 +1525,30 @@ public func matmul<Scalar: Numeric>(
 ) -> (Tensor<Scalar>, (Tensor<Scalar>) -> (Tensor<Scalar>, Tensor<Scalar>)) {
-    let value = matmul(lhs, rhs, transposeA: transposeA, transposeB: transposeB)
+    let value = matmul(lhs, transposed: transposeA, rhs, transposed: transposeB)
     return (value, { v in
```
Use 4-space indentation.

rxwei commented on Sources/TensorFlow/Operators/Math.swift:
```diff
@@ -1505,25 +1505,57 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
-        (matmul(v, rhs.transposed()), matmul(lhs.transposed(), v))
+    let value = matmul(lhs, transposed: transposeA, rhs, transposed: transposeB)
+    return (value, { v in
+      let (lhsGrad, rhsGrad) = { () -> (Tensor<Scalar>, Tensor<Scalar>) in
```
Why does this need to be a closure?

You can declare `let (lhsGrad, rhsGrad): (Tensor<Scalar>, Tensor<Scalar>)` here and assign to it in the switch statement.

eaplatanios commented:
I updated this. I'll update the other PRs that depend on this after it's merged so they use the final version.

eaplatanios commented:
The tests will also not pass for this until the new toolchain is ready.

rxwei commented:
A new toolchain is ready, but it looks like Kokoro is no longer responding.

eaplatanios commented:
@rxwei yeah I also tried to enable it on #141 and it doesn't seem to work.

eaplatanios commented:
Could the `kokoro:force-run` flag possibly work?

rxwei commented:
Looks like it doesn't work. I'll be the CI machine manually tonight.

eaplatanios commented:
Thanks a lot Richard! :)

I think the best way is to go through them one by one and after each is merged I'll go ahead and merge changes with the next and notify you (because there may be some conflicts).

rxwei commented:
Yeah, I'm starting with this one. Building a toolchain right now.

eaplatanios commented:
Sounds good. This, #141, and #142 should be independent of each other and can thus be tested and merged separately. They all pass locally for me.

rxwei commented:
All tests passed. Thank you!
